core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java

```diff
@@ -0,0 +1,171 @@
+                }
+            }
+            return null;
+        }
+    }
+
+    public void start(long startTime, Score_ startingScore) {
+        resetGracePeriod(startTime, startingScore);
+    }
+
+    public void step(long time, Score_ bestScore) {
+        scoresByTime.put(time, bestScore);
+    }
+
+    private void resetGracePeriod(long currentTime, Score_ startingScore) {
+        gracePeriodStartTimeMillis = currentTime;
+        isGracePeriodActive = true;
+
+        // Remove all entries in the map since grace is reset
+        scoresByTime.clear();
+
+        // Put the current best score as the first entry
+        scoresByTime.put(currentTime, startingScore);
+    }
+
+    public boolean isTerminated(long currentTime, Score_ endScore) {
+        if (isGracePeriodActive) {
+            // first score in scoresByTime = first score in grace period window
+            var endpointDiff = LevelScoreDiff.between(scoresByTime.firstEntry().getValue(), endScore);
+            var timeElapsedMillis = currentTime - gracePeriodStartTimeMillis;
+            if (endpointDiff != null && endpointDiff.harderScoreChanged) {
+                // A harder score changed, so reset the grace period
+                resetGracePeriod(currentTime, endScore);
+                return false;
+            }
+            if (timeElapsedMillis >= gracePeriodMillis) {
+                // grace period over, record the reference diff
+                isGracePeriodActive = false;
+                gracePeriodScoreDiff = endpointDiff;
+                return endpointDiff == null;
+            }
+            return false;
+        }
+
+        var startScoreEntry = scoresByTime.floorEntry(currentTime - gracePeriodMillis);
+        var startScore = startScoreEntry.getValue();
+        scoresByTime.subMap(0L, startScoreEntry.getKey()).clear();
+        var scoreDiff = LevelScoreDiff.between(startScore, endScore);
+
+        if (scoreDiff == null) {
+            // no change after grace period, terminate
+            return true;
+        }
+
+        if (scoreDiff.harderScoreChanged) {
+            // A harder score changed, so reset the grace period
+            resetGracePeriod(currentTime, endScore);
+            return false;
+        }
+
+        return scoreDiff.softestScoreDiff / gracePeriodScoreDiff.softestScoreDiff < minimumImprovementRatio;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean isSolverTerminated(SolverScope<Solution_> solverScope) {
+        return isTerminated(System.currentTimeMillis(), (Score_) solverScope.getBestScore());
+    }
+
+    @Override
+    public boolean isPhaseTerminated(AbstractPhaseScope<Solution_> phaseScope) {
+        return isTerminated(System.currentTimeMillis(), phaseScope.getBestScore());
+    }
+
+    @Override
+    public double calculateSolverTimeGradient(SolverScope<Solution_> solverScope) {
+        return -1.0;
```

Comment:
I wonder if there's a way to estimate this based on how close we are to the minimum necessary improvement.